src/aio.rs

```diff
@@ -577,3 +577,91 @@ impl ConnectionLike for MultiplexedConnection {
         self.db
     }
 }
+
+#[cfg(feature = "tokio-rt-core")]
+mod connection_manager {
+    use super::*;
+
+    /// A `ConnectionManager` is a proxy that wraps a multiplexed connection and
+    /// automatically reconnects to the server when necessary.
+    pub struct ConnectionManager {
+        connection_info: ConnectionInfo,
+        connection: MultiplexedConnection,
+    }
+
+    #[cfg(feature = "tokio-rt-core")]
+    impl ConnectionManager {
+        /// Connect to the server and store the connection inside the returned `ConnectionManager`.
+        ///
+        /// TODO: Add `new_tokio`?
+        pub async fn new(connection_info: ConnectionInfo) -> RedisResult<Self> {
+            let (connection, driver) = MultiplexedConnection::new(connect(&connection_info).await?);
+            tokio::spawn(driver);
+            Ok(Self {
+                connection_info,
+                connection,
+            })
+        }
+
+        async fn reconnect(&mut self) -> RedisResult<&mut Self> {
+            let (new_connection, driver) = MultiplexedConnection::new(connect(&self.connection_info).await?);
+            tokio::spawn(driver);
+            let _old_connection = std::mem::replace(&mut self.connection, new_connection);
+            Ok(self)
+        }
+    }
+
+    #[cfg(feature = "tokio-rt-core")]
+    impl ConnectionLike for ConnectionManager {
+        fn req_packed_command<'a>(&'a mut self, cmd: &'a Cmd) -> RedisFuture<'a, Value> {
+            (async move {
+                let result = self.connection.req_packed_command(cmd).await;
+                if let Err(ref e) = result {
+                    if e.is_connection_dropped() {
+                        // TODO: This could lead to multiple reconnection
+                        // attempts being done in parallel... We should
+                        // probably avoid that.
+                        println!("Connection lost, reconnecting");
+                        if let Err(reconnect_failed) = self.reconnect().await {
+                            println!("Reconnecting failed: {}", reconnect_failed);
+                        };
```

Comment:
The current logic will trigger a reconnect attempt once after every "connection dropped" error.

It means that right now multiple reconnection attempts could happen in parallel if multiple errors happen simultaneously (which is very likely).

Does someone have an idea on how to solve that cleanly?

Reply:
Store something that uniquely identifies each individual connection and then when calling `reconnect` you check if the connection is the same or has changed. If it has changed then a concurrent request has already started the reconnection and nothing needs to be done.

Could be an integer incremented on every reconnection attempt or you could store the connection behind an `Arc` and compare the pointers.

Reply:
Ah, @Marwes mentioned `future::Shared` here: https://docs.rs/futures/0.3.1/futures/future/trait.FutureExt.html#method.shared That could be useful.